### PR TITLE
mbbsd: Ensure setting up contact-email

### PR DESCRIPTION
**Pull request by chhsiao1981**
contact-email will be frequently used
for important operations (resetting password / notifications)

We would like to ensure that valid contact-email
exists right after successful login.

Return err in change_contact_email.
The program exits with err if unable to set contact-email
right after successful login.

### CODE REVIEW

**Comment by chhsiao1981** [comment]
#113

File: mbbsd/register.c [hunk 1]
```diff
@@ -1158,6 +1158,18 @@ void
 check_register(void)
 {
     char fn[PATHLEN];
+    int ret = 0;
+
+    // �T�{�@�w�n���p���H�c
+    sprintf(fn, "etc/ensurecontactemail");
+    if(dashf(fn) && !is_valid_email(cuser.email)) {
```
Comment (wens) [inline mbbsd/register.c]: This should be determined at compile time, not based on whether some file exists or not, which introduces system call overhead on each login.

This is no longer your old session based BBS. Many apps log in and out rapidly.

File: include/proto.h [hunk 2]
```diff
@@ -151,7 +151,7 @@ const char *banned_msg(const char *bname);
 
 /* cal */
 const char* money_level(int money);
-int pay(int money, const char *item, ...) GCC_CHECK_FORMAT(2,3); 
+int pay(int money, const char *item, ...) GCC_CHECK_FORMAT(2,3);
```
Comment (wens) [inline include/proto.h]: Please drop all unrelated white space changes.

File: mbbsd/register.c [hunk 3]
```diff
@@ -1158,6 +1158,18 @@ void
...
+        more(fn, NA);
+        pressanykey();
+        ret = change_contact_email();
+        if(ret) {
+          exit(1);
+        }
```
Comment (wens) [inline mbbsd/register.c]: Why not just do a while loop until it succeeds?

**Review by wens — CHANGES_REQUESTED** [review]

File: mbbsd/register.c [hunk 4]
```diff
@@ -1158,6 +1158,16 @@ void
 check_register(void)
 {
     char fn[PATHLEN];
+    int ret = 0;
+
+    // �T�{�@�w�n���p���H�c
+    if(!is_valid_email(cuser.email)) {
+        more("etc/ensurecontactemail\0", NA);
```
Comment (wens) [inline mbbsd/register.c]: C strings are already NUL terminated. You don't need to add the NUL.

File: mbbsd/register.c [hunk 5]
```diff
@@ -1158,6 +1158,16 @@ void
...
+        pressanykey();
+
+        // additional parentheses around "ret =" to please compiler
+        while((ret = change_contact_email()));
```
Comment (wens) [inline mbbsd/register.c]: Space between `while` and parentheses.

**Comment by chhsiao1981** [comment]
Hi @wens～

Thanks for the comments.
I've updated the code accordingly.

**Comment by chhsiao1981** [comment]
We don't need #122 after changing the strategy by looking at the lastlogin.

#114 needs this PR to ensure valid contact email.

File: mbbsd/mbbsd.c [hunk 6]
```diff
@@ -1192,6 +1194,22 @@ user_login(void)
     setproctitle("%s: %s", margs, cuser.userid);
 #endif
 
+    // ensuring the contact email if lastlogin before sometime.
+#ifdef USEREC_EMAIL_IS_CONTACT
+    #ifdef ENSURE_CONTACT_EMAIL_LASTLOGIN
+    if(cuser.lastlogin < ENSURE_CONTACT_EMAIL_LASTLOGIN) {
```
Comment (wens) [inline mbbsd/mbbsd.c]: No. The lastlogin check should be a separate commit.

There should be two separate commits:
a) Force user to add their contact email if the current one is invalid.
b) Force user to update their contact email if their `lastlogin` is before some timestamp.
The second item is optional and should be guarded with `#ifdef FORCE_UPDATE_CONTACT_EMAIL_LASTLOGIN`

File: mbbsd/mbbsd.c [hunk 7]
```diff
@@ -1192,6 +1194,22 @@ user_login(void)
     setproctitle("%s: %s", margs, cuser.userid);
 #endif
 
+    // ensuring the contact email if lastlogin before sometime.
+#ifdef USEREC_EMAIL_IS_CONTACT
+    #ifdef ENSURE_CONTACT_EMAIL_LASTLOGIN
```
Comment (wens) [inline mbbsd/mbbsd.c]: `FORCE_UPDATE_CONTACT_EMAIL_LASTLOGIN` is better.

File: mbbsd/mbbsd.c [hunk 8]
```diff
@@ -1192,6 +1194,22 @@ user_login(void)
...
+    if(cuser.lastlogin < ENSURE_CONTACT_EMAIL_LASTLOGIN) {
+        more("etc/ensurecontactemail", NA);
+        pressanykey();
+
+        pwcuSetEmail(empty_email);
```
Comment (wens) [inline mbbsd/mbbsd.c]: This isn't ideal though. If the user disconnects, they will be left with empty contact email.
You could change the underlying functions to accept a boolean parameter that allows the same email to be used. However in that case it should still be re-verified.

File: mbbsd/mbbsd.c (hunk 6 above)
Comment (wens) [reply mbbsd/mbbsd.c]: Basically, you should add the force update stuff after [4b89509](https://github.com/ptt/pttbbs/commit/4b89509731a7adafab84d4d7465eae885b35867f), but with the style in that commit fixed as I mentioned before.

Comment (wens) [reply mbbsd/mbbsd.c]: Space between `if`/`while` and `(` please.

**Comment by chhsiao1981** [comment]
@wens 

Thank you so much for taking time reviewing again.
I'll try as fast as I can to revise what you suggested.

File: mbbsd/mbbsd.c (hunk 6 above)
Comment (chhsiao1981) [reply mbbsd/mbbsd.c]: > Basically, you should add the force update stuff after [4b89509](https://github.com/ptt/pttbbs/commit/4b89509731a7adafab84d4d7465eae885b35867f), but with the style in that commit fixed as I mentioned before.

Sorry for the wrong place of the code.
`lastlogin` is updated in `pwcuLoginSave` and it is after `check_register`
I'll change the code block accordingly.

**Comment by chhsiao1981** [comment]
hi @wens 

Thank you so much for taking time reviewing again.
I believe that I've followed your guideline updating the code.

File: mbbsd/register.c [hunk 9]
```diff
@@ -572,8 +572,25 @@ query_yn(int y, const char *msg)
...
+//     email.
+//     0: not skip checking (to check whether they are the same.)
+//     1: skip checking (not to check whether they are the same.)
+//
+// Return:
+//     0 (REGISTER_OK): ok
+//     < 0: error code
 static int
-register_email_verification(email_input_t *ein)
+register_email_verification(email_input_t *ein, int is_skip_check_same_email)
```
Comment (wens) [inline mbbsd/register.c]: Use `bool`. Parameter should be named `skip_same_email_check`.

File: sample/etc/ensurevalidcontactemail [hunk 10]
```diff
@@ -1,4 +1,4 @@
-�z���p���H�c�w�Q���m�A
+�z���p���H�c���O���Ī��H�c�A
```
Comment (wens) [inline sample/etc/ensurevalidcontactemail:1]: Squash this change, both the text change and rename, into the previous commit.

File: mbbsd/register.c [hunk 11]
```diff
@@ -1712,18 +1737,25 @@ static int notify_email_change(const char *userid, const char *email)
...
+//     whether to skip checking the input email is the same as the original
+//     email.
+//     0: not skip checking (to check whether they are the same.)
+//     1: skip checking (not to check whether they are the same.)
+//
 // Return:
 //   int: 0: ok -1: err
 int
-change_contact_email()
+change_contact_email(int is_skip_check_same_email)
```
Comment (wens) [inline mbbsd/register.c]: Use `bool`. Parameter should be named `skip_same_email_check`.

File: mbbsd/register.c [hunk 12]
```diff
@@ -1162,11 +1179,19 @@ check_register(void)
 
     // �T�{�@�w�n���p���H�c
     if (!is_valid_email(cuser.email)) {
-        more("etc/ensurecontactemail", NA);
+        more("etc/ensurevalidcontactemail", NA);
```
Comment (wens) [inline mbbsd/register.c:1182]: Squash this into the first commit.

File: mbbsd/register.c [hunk 13]
```diff
@@ -1162,11 +1179,19 @@ check_register(void)
...
         pressanykey();
 
         // additional parentheses around "ret =" to please compiler
-        while( (ret = change_contact_email()) );
+        while ((ret = change_contact_email(0)));
+#ifdef FORCE_UPDATE_CONTACT_EMAIL_LASTLOGIN
+    } else if (cuser.lastlogin < FORCE_UPDATE_CONTACT_EMAIL_LASTLOGIN) {
```
Comment (wens) [inline mbbsd/register.c]: `cuser.lastlogin` is already updated [here in `mbbsd.c:user_login()`](https://github.com/ptt/pttbbs/blob/master/mbbsd/mbbsd.c#L1197) before `check_register()` is called in the same function.

Instead, you should use `last_login_time`, which is a backup of `cuser.lastlogin` saved early in `start_client()` before `user_login()` is called.

**Comment by wens** [comment]
The commit subject for the second commit should be different, and should read something like "implement force updating of contact email" or something.

**Review by wens — CHANGES_REQUESTED** [review]

File: mbbsd/register.c (hunk 13 above)
Comment (chhsiao1981) [reply mbbsd/register.c]: It looks like [here in `mbbsd.c:user_login()`](https://github.com/ptt/pttbbs/blob/master/mbbsd/mbbsd.c#L1197)
actually only sets the human-readable format of `cuser.lastlogin` to `lasttime`, but not changes `cuser.lastlogin`
([https://github.com/ptt/pttbbs/blob/master/common/sys/time.c#L153](https://github.com/ptt/pttbbs/blob/master/common/sys/time.c#L153),
 [https://linux.die.net/man/3/localtime_r](https://linux.die.net/man/3/localtime_r)]

However, it looks like [the convention in the code is using `last_login_time`](https://github.com/ptt/pttbbs/blob/master/mbbsd/mbbsd.c#L1269)

I'll follow the convention accordingly.

**Comment by chhsiao1981** [comment]
hi @wens ～

Thank you so much for the review and comments again!
I believe that I've followed all you requests～

Can you review it again？～

Thanks!
